Name : Gourav Verma<br>
Class : DSC540-T303<br>
Topic : Data Cleaning and Transforming<br>
Assignment : Weeks 7 & 8 Excercises<br>

In [1]:
# Import important libraries
import pandas as pd
import numpy as np
from datetime import datetime

In [2]:
met = pd.read_csv('MetObjects.csv', parse_dates=[9])
met.head()

C:\Users\f4puslg\AppData\Local\Continuum\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3058: DtypeWarning: Columns (7,8,10,11,18,27,28,29,30,31,32,33,34,35,36,37,39) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Object Number  Is Highlight  Is Public Domain  Object ID  \
0    1979.486.1         False             False          1   
1    1980.264.5         False             False          2   
2      67.265.9         False             False          3   
3     67.265.10         False             False          4   
4     67.265.11         False             False          5   

                 Department Object Name                         Title Culture  \
0  American Decorative Arts        Coin  One-dollar Liberty Head Coin     NaN   
1  American Decorative Arts        Coin  Ten-dollar Liberty Head Coin     NaN   
2  American Decorative Arts        Coin    Two-and-a-Half Dollar Coin     NaN   
3  American Decorative Arts        Coin    Two-and-a-Half Dollar Coin     NaN   
4  American Decorative Arts        Coin    Two-and-a-Half Dollar Coin     NaN   

  Period Dynasty  ... Subregion Locale Locus Excavation River Classification  \
0    NaN     NaN  ...       NaN    NaN   NaN        NaN   NaN          Metal   
1    NaN     NaN  ...       NaN    NaN   NaN        NaN   NaN          Metal   
2    NaN     NaN  ...       NaN    NaN   NaN        NaN   NaN          Metal   
3    NaN     NaN  ...       NaN    NaN   NaN        NaN   NaN          Metal   
4    NaN     NaN  ...       NaN    NaN   NaN        NaN   NaN          Metal   

  Rights and Reproduction                                     Link Resource  \
0                     NaN  http://www.metmuseum.org/art/collection/search/1   
1                     NaN  http://www.metmuseum.org/art/collection/search/2   
2                     NaN  http://www.metmuseum.org/art/collection/search/3   
3                     NaN  http://www.metmuseum.org/art/collection/search/4   
4                     NaN  http://www.metmuseum.org/art/collection/search/5   

         Metadata Date                                Repository  
0  4/3/2017 8:00:08 AM  Metropolitan Museum of Art, New York, NY  
1  4/3/2017 8:00:08 AM  Metropolitan Museum of Art, New York, NY  
2  4/3/2017 8:00:08 AM  Metropolitan Museum of Art, New York, NY  
3  4/3/2017 8:00:08 AM  Metropolitan Museum of Art, New York, NY  
4  4/3/2017 8:00:08 AM  Metropolitan Museum of Art, New York, NY  

[5 rows x 43 columns]

In [3]:
met.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 448203 entries, 0 to 448202
Data columns (total 43 columns):
Object Number              448203 non-null object
Is Highlight               448203 non-null bool
Is Public Domain           448203 non-null bool
Object ID                  448203 non-null int64
Department                 448203 non-null object
Object Name                445568 non-null object
Title                      416906 non-null object
Culture                    186518 non-null object
Period                     71882 non-null object
Dynasty                    23018 non-null object
Reign                      10817 non-null object
Portfolio                  20370 non-null object
Artist Role                259909 non-null object
Artist Prefix              88928 non-null object
Artist Display Name        261111 non-null object
Artist Display Bio         224064 non-null object
Artist Suffix              10212 non-null object
Artist Alpha Sort          261088 non-null object


In [4]:
met.shape

(448203, 43)

1. **File has 448203 rows and 43 columns, out of that we have 2 boolean, 3 integers and 38 object variables.**
2. **More than half of the variable has missing values.**

### Finding Missing Value

In [5]:
# Finding Missing value counts for each column

met.isna().sum().sort_values(ascending = False)

River                      446100
County                     445715
Locus                      441264
State                      439843
Artist Suffix              437991
Reign                      437386
Locale                     433108
Excavation                 432684
Portfolio                  427833
Subregion                  426487
Rights and Reproduction    425228
Dynasty                    425185
City                       417683
Region                     417125
Geography Type             389740
Period                     376321
Country                    373753
Artist Prefix              359275
Culture                    261685
Artist Nationality         252071
Artist End Date            235378
Artist Begin Date          232969
Artist Display Bio         224139
Artist Role                188294
Artist Alpha Sort          187115
Artist Display Name        187092
Dimensions                  62843
Classification              58279
Title                       31297
Object Date   

In [6]:
# Find missing value percentage
met.isna().mean().sort_values(ascending = False).round(4)*100

River                      99.53
County                     99.44
Locus                      98.45
State                      98.13
Artist Suffix              97.72
Reign                      97.59
Locale                     96.63
Excavation                 96.54
Portfolio                  95.46
Subregion                  95.15
Rights and Reproduction    94.87
Dynasty                    94.86
City                       93.19
Region                     93.07
Geography Type             86.96
Period                     83.96
Country                    83.39
Artist Prefix              80.16
Culture                    58.39
Artist Nationality         56.24
Artist End Date            52.52
Artist Begin Date          51.98
Artist Display Bio         50.01
Artist Role                42.01
Artist Alpha Sort          41.75
Artist Display Name        41.74
Dimensions                 14.02
Classification             13.00
Title                       6.98
Object Date                 3.48
Medium    

### Remove Duplicates

In [7]:
met.duplicated(['Object Number']).sum()

2576

It seems we have 2576 duplicate entries in the file. Lets drop them.

In [8]:
# Remove Duplicates
metd = met.drop_duplicates(subset='Object Number')
metd.shape

(445627, 43)

### Create Hierarchical Index

In [9]:
# Create a small dataframe containing selected columns

metn = metd[['Object Number', 'Department', 'Title', 'Object Name', 'Medium', 'Classification', 'Is Public Domain']]

# Select random sample of 1000 rows
metr = metn.sample(1000)

# Reset Index
metr = metr.reset_index()

In [10]:
# Create Hierarchical indexing, using columns department and medium

meth = metr.set_index(['Department', 'Medium'])
meth.head()

index  \
Department                       Medium                                                  
Drawings and Prints              Engraving; first state of two (New Hollstein)  382913   
Egyptian Art                     Stone                                          374127   
American Paintings and Sculpture Pastel on paper, mounted on cardboard           16552   
Drawings and Prints              Wood engraving                                 210198   
Arms and Armor                   Wood, lacquer, abalone shell, gold              21732   

                                                                                 Object Number  \
Department                       Medium                                                          
Drawings and Prints              Engraving; first state of two (New Hollstein)       59.570.10   
Egyptian Art                     Stone                                              12.181.105   
American Paintings and Sculpture Pastel on paper, mounted on cardboard                  68.200   
Drawings and Prints              Wood engraving                                 1978.643.10(1)   
Arms and Armor                   Wood, lacquer, abalone shell, gold                   2000.405   

                                                                                                                            Title  \
Department                       Medium                                                                                             
Drawings and Prints              Engraving; first state of two (New Hollstein)  Ahab Rending His Clothes Upon Hearing Elijah's...   
Egyptian Art                     Stone                                                                               Cylinder jar   
American Paintings and Sculpture Pastel on paper, mounted on cardboard                                          At the Milliner's   
Drawings and Prints              Wood engraving                                 XIXth Century German Broadsheets--Scenes publi...   
Arms and Armor                   Wood, lacquer, abalone shell, gold                                                 Saddle (Kura)   

                                                                                     Object Name  \
Department                       Medium                                                            
Drawings and Prints              Engraving; first state of two (New Hollstein)             Print   
Egyptian Art                     Stone                                          Jar, cylindrical   
American Paintings and Sculpture Pastel on paper, mounted on cardboard                   Drawing   
Drawings and Prints              Wood engraving                                            Print   
Arms and Armor                   Wood, lacquer, abalone shell, gold                Saddle (Kura)   

                                                                                              Classification  \
Department                       Medium                                                                        
Drawings and Prints              Engraving; first state of two (New Hollstein)                        Prints   
Egyptian Art                     Stone                                                                   NaN   
American Paintings and Sculpture Pastel on paper, mounted on cardboard                              Drawings   
Drawings and Prints              Wood engraving                                                       Prints   
Arms and Armor                   Wood, lacquer, abalone shell, gold             Equestrian Equipment-Saddles   

                                                                                Is Public Domain  
Department                       Medium                                                           
Drawings and Prints              Engraving; first state of two (New Hollstein)             False  
Egyptian Art                     Stone          

### Reshaping

In [11]:
# Reshaping the sample dataframe. Changing columns into rows.
st = metr.stack()
st

0    index                                                          382913
     Object Number                                               59.570.10
     Department                                        Drawings and Prints
     Title               Ahab Rending His Clothes Upon Hearing Elijah's...
     Object Name                                                     Print
                                               ...                        
999  Title                                       Glass trick bottle or cup
     Object Name                                       Trick bottle or cup
     Medium                                                          Glass
     Classification                                                  Glass
     Is Public Domain                                                 True
Length: 7776, dtype: object

In [12]:
# Reshape from the rows into the columns
st.unstack()

index   Object Number                        Department  \
0    382913       59.570.10               Drawings and Prints   
1    374127      12.181.105                      Egyptian Art   
2     16552          68.200  American Paintings and Sculpture   
3    210198  1978.643.10(1)               Drawings and Prints   
4     21732        2000.405                    Arms and Armor   
..      ...             ...                               ...   
995  209509         40.91.9               Drawings and Prints   
996    1363      39.127a, b          American Decorative Arts   
997   74002     1978.582.56                 Costume Institute   
998  342890          59.160       Modern and Contemporary Art   
999  142170       74.51.150               Greek and Roman Art   

                                                 Title          Object Name  \
0    Ahab Rending His Clothes Upon Hearing Elijah's...                Print   
1                                         Cylinder jar     Jar, cylindrical   
2                                    At the Milliner's              Drawing   
3    XIXth Century German Broadsheets--Scenes publi...                Print   
4                                        Saddle (Kura)        Saddle (Kura)   
..                                                 ...                  ...   
995                                             Margit              Drawing   
996                                             Caster               Caster   
997                                              Shirt                Shirt   
998                   Homage to the Square:  With Rays             Painting   
999                          Glass trick bottle or cup  Trick bottle or cup   

                                            Medium  \
0    Engraving; first state of two (New Hollstein)   
1                                            Stone   
2            Pastel on paper, mounted on cardboard   
3                                   Wood engraving   
4               Wood, lacquer, abalone shell, gold   
..                                             ...   
995                              Pen and black ink   
996                                         Silver   
997                                         cotton   
998                                Oil on Masonite   
999                                          Glass   

                   Classification Is Public Domain  
0                          Prints            False  
1                             NaN            False  
2                        Drawings             True  
3                          Prints            False  
4    Equestrian Equipment-Saddles             True  
..                            ...              ...  
995                      Drawings             True  
996                        Silver             True  
997                           NaN            False  
998                     Paintings            False  
999                         Glass             True  

[1000 rows x 8 columns]

### Grouping with Functions

In [13]:
# Lets find whats the count if the object is public domain or not.

metg = metr.set_index(['Is Public Domain'])
metg.groupby(bool).count()

index  Object Number  Department  Title  Object Name  Medium  \
False    550            550         550    534          545     538   
True     450            450         450    400          449     446   

       Classification  
False             462  
True              402

### Crosstab

In [14]:
# Summarize the total records by calssifications and departments, and display top 10 classifications

met_ic = pd.crosstab(metn.Classification, metn.Department, margins = True)
met_ic.sort_values('All', ascending=False).head(11)

Department            American Decorative Arts  \
Classification                                   
All                                      12330   
Prints                                      18   
Prints|Ephemera                              0   
Photographs                                100   
Drawings                                   538   
Books                                        0   
Ceramics                                  2606   
Textiles-Woven                               0   
Photographs|Ephemera                         0   
Paintings                                  161   
Glass                                     2650   

Department            American Paintings and Sculpture  \
Classification                                           
All                                               5451   
Prints                                              17   
Prints|Ephemera                                      0   
Photographs                                          2   
Drawings                                          2804   
Books                                                0   
Ceramics                                             1   
Textiles-Woven                                       0   
Photographs|Ephemera                                 0   
Paintings                                         1836   
Glass                                                0   

Department            Ancient Near Eastern Art  Arms and Armor  \
Classification                                                   
All                                       6185           13484   
Prints                                       0               0   
Prints|Ephemera                              0               0   
Photographs                                  0               6   
Drawings                                     0               2   
Books                                        0               0   
Ceramics                                    27               0   
Textiles-Woven                              26               0   
Photographs|Ephemera                         0               0   
Paintings                                    0               4   
Glass                                        0               0   

Department            Arts of Africa, Oceania, and the Americas  Asian Art  \
Classification                                                               
All                                                       12423      36727   
Prints                                                        4       4858   
Prints|Ephemera                                               0          0   
Photographs                                                 131         46   
Drawings                                                      0          0   
Books                                                         2          0   
Ceramics                                                      3       6315   
Textiles-Woven                                             1260       4253   
Photographs|Ephemera                                          0          0   
Paintings                                                     1       4260   
Glass                                                         1         78   

Department            Costume Institute  Drawings and Prints  Egyptian Art  \
Classification                                                               
All                                   5               152893          3041   
Prints                                0                61735             0   
Prints|Ephemera                       0                30032             0   
Photographs                           0                  150             0   
Drawings                              0                15909             0   
Books                                 0                13339             0   
Ceramics                              0                    0             0   
Textiles-Woven    

In [15]:
type(metd['Object Begin Date'])

pandas.core.series.Series

In [16]:
pd.to_datetime(metd['Object Begin Date'])

0        1970-01-01 00:00:00.000001853
1        1970-01-01 00:00:00.000001901
2        1970-01-01 00:00:00.000001909
3        1970-01-01 00:00:00.000001909
4        1970-01-01 00:00:00.000001909
                      ...             
448198   1970-01-01 00:00:00.000001545
448199   1970-01-01 00:00:00.000001624
448200   1970-01-01 00:00:00.000001595
448201   1970-01-01 00:00:00.000001715
448202   1970-01-01 00:00:00.000001788
Name: Object Begin Date, Length: 445627, dtype: datetime64[ns]

In [17]:
obd = str(metd['Object Begin Date'])

In [18]:
datetime.strptime(obd, '%Y')

ValueError: time data '0         1853\n1         1901\n2         1909\n3         1909\n4         1909\n          ... \n448198    1545\n448199    1624\n448200    1595\n448201    1715\n448202    1788\nName: Object Begin Date, Length: 445627, dtype: int64' does not match format '%Y'

In [ ]:
metd['Artist Begin Date'].dt.year

### TimeSeries

I tried several option to convert my dataframe's series into datetime format, but got no success. 
Also the data in the date fields of file is not accurate. I see negative values. 
Need more analysis and work on the data.